README.md
# Malcontent
[![Build Status](https://travis-ci.org/ecmendenhall/malcontent.png)](https://travis-ci.org/ecmendenhall/malcontent)

Content Security Policy is a new browser security mechanism to prevent
[cross site
scripting](https://www.owasp.org/index.php/Top_10_2013-A3-Cross-Site_Scripting_(XSS\))
attacks. By sending a 'Content-Security-Policy' header in HTTP
responses, web applications can provide rules and restrictions for
client side scripts, plugins, frames, and other resources. Most [modern browsers](http://caniuse.com/contentsecuritypolicy)
enforce these restrictions, preventing injection of malicious
javascript, and effectively shutting down most XSS attacks. 

Malcontent is a simple middleware handler for adding the CSP header to Ring applications. Simply specify a security policy as a Clojure map and malcontent will include it in responses to supported browsers.

## Usage
Malcontent looks for a policy file at `config/security_policy.clj` by default (but you can pass in your own path if you'd prefer). Here are the examples from the [HTML5 Rocks introduction to CSP](http://www.html5rocks.com/en/tutorials/security/content-security-policy/#real-world-usage) as malcontent maps:

__Social media widgets:__

```clj
{:sources {:script ["https://apis.google.com"
                    "https://platform.twitter.com"]
           :frame  ["https://plusone.google.com"
                    "https://facebook.com"
                    "https://platform.twitter.com"]}}
```

__Lockdown:__

```clj
{:sources {:default :none
           :script  "https://cdn.mybank.net"
           :style   "https://cdn.mybank.net"
           :img     "https://cdn.mybank.net"
           :connect "https://api.mybank.com"
           :frame   :self}}
```

__SSL Only__:

```clj
{:sources {:default "https:"
           :script  ["https:" :unsafe-inline]
           :style   ["https:" :unsafe-inline]}}
```

In addition to sources, policy maps may include the `:sandbox` and
`:report-uri` directives. Sources may be specified as strings,
keywords, or vectors. Here's an example with every directive enabled:

```clj
{:sources {:default :self
           :script ["https://trustedscripts.com" :unsafe-eval]
           :style ["*.styles.example.com"
                   "https://inlinestyles.info"
                   :unsafe-inline]
           :img "*"
           :connect "https:"
           :font "http://webfonts.biz"
           :object :none
           :media ["http://media.example.com"]
           :frame :self}
 :sandbox [:allow-forms
           :allow-scripts
           :allow-top-navigation]
 :report-uri "/some/report-uri"}
```

Mapping [CSP directives](http://content-security-policy.com/) to Clojure values is simple: special values like `'self'`, `'none'`, and `'unsafe-inline'` become keywords, while sources like `https:` `*.example.com` and `https://example.com` become strings.

To include the policy in outgoing responses, just include `add-content-security-policy` as a middleware wrapper:

```clj
(ns my-great-webapp.core
  (:require [malcontent.core :refer [add-content-security-policy]]))

(defroutes app-routes
  (GET "/" [] my-great-request-handler))
  
(def app (-> routes
             (add-content-security-policy)))
```

If you'd prefer to load the policy from somewhere besides the default location, pass it as a keyword argument:

```clj
(def app (-> routes
             (add-content-security-policy :config-path "resources/policy.clj")))
```

For help writing a good security policy, check out the resources below.
                 
## Installation
Include malcontent as a dependency in `project.clj`:

```clj
[malcontent "0.2.0-SNAPSHOT"]
```

## Content Security Policy Resources
- [CSP Playground](http://www.cspplayground.com/) &mdash; interactive examples of common CSP violations and a policy validator.
- [HTML5 Rocks introduction to CSP](http://www.html5rocks.com/en/tutorials/security/content-security-policy/) &mdash; a helpful overview of CSP use cases.
- [CSP quick reference](http://content-security-policy.com/) &mdash; a CSP directive cheat sheet.
- [Cross-browser support](http://caniuse.com/contentsecuritypolicy) &mdash; the current state of cross-browser CSP support.
- [Policy recommendation bookmarklet](http://brandon.sternefamily.net/posts/2010/10/content-security-policy-recommendation-bookmarklet/) &mdash; a one-click tool that analyzes existing pages and generates a policy recommendation.
- [Automated policy generator](http://cspisawesome.com/) &mdash; create a valid policy online by clicking helpful checkboxes.
- [W3C CSP 1.0 specification](http://www.w3.org/TR/CSP/) &mdash; straight from the source.

## More to come...
- Map custom policies to individual routes.
- Add a default logger for policy violations.

## License
Malcontent is released under the Apache License, v2.0. For details, see the file `LICENSE.md` in this repository.


CHANGELOG.md
# Changelog

## 0.2.1 - 2022-06-09

## 0.2.0 - 2022-06-09

### Removed
- SNAPSHOT project; lint; upgrade and remove unused `ring` dependency [#3](https://github.com/droitfintech/malcontent/pull/3)

## 0.1.0 - 2019-03-08

### Added
- Handle IE 11 [#1](https://github.com/droitfintech/malcontent/pull/1)


spec/malcontent/policy_spec.clj
(ns malcontent.policy-spec
  (:require
   [clojure.string :as str]
   [malcontent.policy :refer [get-directives get-sources load-policy
                              make-directive make-policy make-source-directive
                              map-join source-values to-source]]
   [speclj.core :refer [describe it should should=]]))

(def clj-policy {:sources {:script ["https://apis.google.com"
                                    "https://platform.twitter.com"]
                           :frame "https://plusone.google.com"}
                 :sandbox :allow-forms
                 :report-uri "/some-report-uri"})

(def source-policy (str "frame-src https://plusone.google.com; "
                        "script-src https://apis.google.com "
                        "https://platform.twitter.com"))

(def simple-http-policy (str source-policy "; "
                             "sandbox 'allow-forms'; "
                             "report-uri /some-report-uri"))

(def tricky-clj-policy {:sources {:style ["*.styles.example.com"
                                          "https://inlinestyles.info"
                                          :unsafe-inline]
                                  :connect "https:"
                                  :script ["https://trustedscripts.com"
                                           :unsafe-eval]
                                  :img "*"
                                  :media ["http://media.example.com"]
                                  :frame :self}
                        :sandbox [:allow-forms
                                  :allow-scripts
                                  :allow-top-navigation]
                        :report-uri "/some/report-uri"})

(def tricky-http-policy (str "img-src *; "
                             "style-src *.styles.example.com "
                             "https://inlinestyles.info "
                             "'unsafe-inline'; "
                             "media-src http://media.example.com; "
                             "connect-src https:; "
                             "frame-src 'self'; "
                             "script-src https://trustedscripts.com " 
                             "'unsafe-eval'; "
                             "sandbox 'allow-forms' 'allow-scripts' "
                             "'allow-top-navigation'; "
                             "report-uri /some/report-uri"))

(defn- policy=
  "Checks two policy strings `expected` and `actual` are equal, irrespective of
  the order."
  [expected actual]
  (letfn [(normalize [s]
            (into #{} (map str/trim) (str/split s #";")))]
    (= (normalize expected) (normalize actual))))

(describe "Constructing policies"

          (it "converts keywords to source values"
              (should= "'none'"
                       (to-source :none)))

          (it "does not convert strings"
              (should= "https://apis.google.com"
                       (to-source "https://apis.google.com")))

          (it "maps over a collection, then joins the results"
              (should= "2~*3~*4~*5"
                       (map-join inc "~*" [1 2 3 4])))
          
          (it "converts a list of sources into a string"
              (should= "https://apis.google.com https://platform.twitter.com"
                       (source-values ((clj-policy :sources) :script))))

          (it "converts a single source into a string"
              (should= "https://apis.google.com"
                       (source-values "https://apis.google.com")))
          
          (it "converts a keyword/source pair into a directive"
              (should= "frame-src https://plusone.google.com"
                       (make-source-directive [:frame "https://plusone.google.com"])))

          (it "converts a simple keyword/value pair into a directive"
              (should= "sandbox 'allow-forms'"
                       (make-directive [:sandbox :allow-forms])))
          
          (it "converts the sources map into a policy string"
              (should (policy= source-policy
                               (get-sources (clj-policy :sources)))))

          (it "converts the remaining policy into a policy string"
              (should (policy= "sandbox 'allow-forms'; report-uri /some-report-uri"
                               (get-directives (dissoc clj-policy :sources)))))

          (it "converts a full policy map into a policy string"
              (should (policy= simple-http-policy
                               (make-policy clj-policy))))

          (it "converts a policy map without sources into a valid policy string"
              (should (policy= "sandbox 'allow-forms'; report-uri /some-report-uri"
                               (make-policy (dissoc clj-policy :sources)))))

          (it "converts a policy map without directives into a valid policy string"
              (should (policy= (get-sources (clj-policy :sources))
                               (make-policy (dissoc clj-policy :sandbox :report-uri)))))

          (it "converts policy maps with special directives into valid policy strings"
              (should (policy= tricky-http-policy
                               (make-policy tricky-clj-policy))))

          (it "loads policy maps from the default location"
              (should (policy= tricky-http-policy
                               (make-policy (load-policy))))))


spec/malcontent/browser_spec.clj
(ns malcontent.browser-spec
  (:require [speclj.core :refer [describe it should should=]]
            [malcontent.browser :refer [gecko-header? get-user-agent parse-user-agent-string select-header standard-header? webkit-header?]]))

(def empty-headers {:headers {}})

(def user-agent-string (str "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_8_4) "
                            "AppleWebKit/537.36 (KHTML, like Gecko) "
                            "Chrome/28.0.1500.95 Safari/537.36"))

(def mock-request  {:headers {"user-agent" user-agent-string}})

(def chrome  {:version [28 0 1500 95] :name "Chrome" :family "CHROME"})
(def firefox {:version [23] :name "Firefox" :family "FIREFOX"})
(def opera   {:version [16] :name "Opera" :family "OPERA"})
(def safari  {:version [5] :name "Safari" :family "SAFARI"})

(def old-chrome  {:version [25 0 1500 95] :name "Chrome" :family "CHROME"})
(def old-firefox {:version [23] :name "Firefox" :family "FIREFOX"})
(def internet-explorer {:version [10] :name "Internet Explorer" :family "IE"})

(def standard-strings [
  "Mozilla/5.0 (Windows NT 6.2; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/29.0.1547.2 Safari/537.36"
  "Mozilla/5.0 (Windows NT 6.2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/28.0.1467.0 Safari/537.36"
  "Mozilla/5.0 (X11; CrOS i686 3912.101.0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/27.0.1453.116 Safari/537.36"
  "Mozilla/5.0 (Windows NT 6.1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/27.0.1453.93 Safari/537.36"
  "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_8_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/27.0.1453.93 Safari/537.36"
  "Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:23.0) Gecko/20131011 Firefox/23.0"
  "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:23.0) Gecko/20130406 Firefox/23.0"
  "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.8; rv:24.0) Gecko/20100101 Firefox/24.0"])

(def webkit-strings [
  "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.17 (KHTML, like Gecko) Chrome/24.0.1312.60 Safari/537.17"
  "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_8_2) AppleWebKit/537.17 (KHTML, like Gecko) Chrome/24.0.1309.0 Safari/537.17"
  "Mozilla/5.0 (Windows NT 6.2; WOW64) AppleWebKit/537.15 (KHTML, like Gecko) Chrome/24.0.1295.0 Safari/537.15"
  "Mozilla/5.0 (Windows NT 6.2; WOW64) AppleWebKit/537.14 (KHTML, like Gecko) Chrome/24.0.1292.0 Safari/537.14"
  "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_6_0) AppleWebKit/537.4 (KHTML, like Gecko) Chrome/22.0.1229.79 Safari/537.4"
  "Mozilla/5.0 (Windows NT 6.2; WOW64) AppleWebKit/537.1 (KHTML, like Gecko) Chrome/19.77.34.5 Safari/537.1"
  "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/536.3 (KHTML, like Gecko) Chrome/19.0.1062.0 Safari/536.3"
  "Mozilla/5.0 (Windows; U; Windows NT 6.1; sv-SE) AppleWebKit/533.19.4 (KHTML, like Gecko) Version/5.0.3 Safari/533.19.4"
  "Mozilla/5.0 (Windows; U; Windows NT 6.1; fr-FR) AppleWebKit/533.20.25 (KHTML, like Gecko) Version/5.0.4 Safari/533.20.27"
  "Mozilla/5.0 (Macintosh; U; Intel Mac OS X 10_6_8; de-at) AppleWebKit/533.21.1 (KHTML, like Gecko) Version/5.0.5 Safari/533.21.1"
  "Mozilla/5.0 (iPad; CPU OS 5_1 like Mac OS X) AppleWebKit/534.46 (KHTML, like Gecko ) Version/5.1 Mobile/9B176 Safari/7534.48.3"
  "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_7_3) AppleWebKit/534.55.3 (KHTML, like Gecko) Version/5.1.3 Safari/534.53.10"
  "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_6_8) AppleWebKit/537.13+ (KHTML, like Gecko) Version/5.1.7 Safari/534.57.2"
  "Mozilla/5.0 (iPad; CPU OS 6_0 like Mac OS X) AppleWebKit/536.26 (KHTML, like Gecko) Version/6.0 Mobile/10A5355d Safari/8536.25"
  ])

(def gecko-strings [
  "Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:22.0) Gecko/20130328 Firefox/22.0"
  "Mozilla/5.0 (Windows NT 6.1; rv:22.0) Gecko/20130405 Firefox/22.0"
  "Mozilla/5.0 (Windows NT 6.2; Win64; x64; rv:16.0.1) Gecko/20121011 Firefox/21.0.1"
  "Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:16.0.1) Gecko/20121011 Firefox/21.0.1"
  "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:21.0) Gecko/20130331 Firefox/21.0"
  "Mozilla/5.0 (X11; Linux i686; rv:21.0) Gecko/20100101 Firefox/21.0"
  "Mozilla/5.0 (Windows NT 5.1; rv:21.0) Gecko/20100101 Firefox/21.0"
  "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.8; rv:21.0) Gecko/20100101 Firefox/21.0"
  "Mozilla/5.0 (Windows NT 6.1; rv:6.0) Gecko/20100101 Firefox/19.0"
  "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:18.0)  Gecko/20100101 Firefox/18.0"
  "Mozilla/5.0 (Windows NT 6.2; Win64; x64; rv:16.0.1) Gecko/20121011 Firefox/16.0.1"
  "Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:5.0) Gecko/20110619 Firefox/5.0"
  "Mozilla/5.0 (X11; Linux x86_64; rv:2.2a1pre) Gecko/20100101 Firefox/4.2a1pre"
  "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.6; rv:2.0b8) Gecko/20100101 Firefox/4.0b8"
  "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:2.0b11pre) Gecko/20110128 Firefox/4.0b11pre"
  "Mozilla/5.0 (compatible; MSIE 10.0; Windows NT 6.1; WOW64; Trident/6.0)"
  "Mozilla/5.0 (compatible; MSIE 10.0; Windows NT 6.1; Trident/5.0)"
  "Mozilla/5.0 (compatible; MSIE 10.0; Macintosh; Intel Mac OS X 10_7_3; Trident/6.0)"
  "Mozilla/5.0 (compatible; MSIE 10.6; Windows NT 6.1; Trident/5.0; InfoPath.2; SLCC1; .NET CLR 3.0.4506.2152; .NET CLR 3.5.30729; .NET CLR 2.0.50727) 3gpp-gba UNTRUSTED/1.0"])

(defn correct-header? [predicate collection & options]
  (every? (fn [ua-string]
            (when options
              (when (options :verbose) (println "Testing user agent: " ua-string)))
            (predicate (parse-user-agent-string ua-string)))
          collection))

(describe "Examining requests"

          (it "should parse a user-agent string into a map"
              (should= chrome
                       (parse-user-agent-string user-agent-string)))

          (it "should extract the user-agent from the request headers"
              (should= chrome
                       (get-user-agent mock-request)))

          (it "should identify browsers that implement the standard header"
              (should (standard-header? chrome))
              (should (standard-header? firefox))
              (should (standard-header? opera)))

          (it "should identify browsers that implement the webkit header"
              (should (webkit-header? old-chrome))
              (should (webkit-header? safari)))

          (it "should identify browsers that implement the gecko header"
              (should (gecko-header? old-firefox))
              (should (gecko-header? internet-explorer)))

          (it "should select the correct header from the user-agent string"
              (should (correct-header? standard-header? standard-strings))
              (should (correct-header? webkit-header? webkit-strings))
              (should (correct-header? gecko-header? gecko-strings)))
          
          (it "should select the correct header from the request"
              (should= "Content-Security-Policy"
                       (select-header mock-request)))) 


spec/malcontent/middleware_spec.clj
(ns malcontent.middleware-spec
  (:require [speclj.core :refer [describe it should=]]
            [malcontent.middleware :refer [add-content-security-policy]]
            [malcontent.policy-spec :refer [tricky-http-policy]]))

(def user-agent-string (str "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_8_4) "
                            "AppleWebKit/537.36 (KHTML, like Gecko) "
                            "Chrome/28.0.1500.95 Safari/537.36"))

(def mock-request  {:headers {"user-agent" user-agent-string}})

(describe "Wrapping responses"

          (it "should add a Content-Security-Policy header to outgoing responses"
              (should= {:headers {"Content-Security-Policy" tricky-http-policy
                                  "user-agent" user-agent-string}}
                       ((add-content-security-policy identity) mock-request)))
          (it "should load policy configurations from a custom location when passed as a keyword arg"
              (should= {:headers {"Content-Security-Policy" "default-src 'self'"
                                  "user-agent" user-agent-string}}
                       ((add-content-security-policy identity :config-path "config/alternate_policy.clj") mock-request))))


src/malcontent/browser.clj
(ns malcontent.browser
  (:import
   (net.sf.uadetector.service UADetectorServiceFactory)))

(defn parse-user-agent-string [user-agent-string]
  (let [parser (UADetectorServiceFactory/getResourceModuleParser)
        user-agent (.parse parser user-agent-string)]
    (-> {}
        (assoc :family (str (.getFamily user-agent)))
        (assoc :name (.getName user-agent))
        (assoc :version (vec (map (fn [number-str]
                                    (try (Integer/parseInt number-str)
                                         (catch NumberFormatException _ 0)))
                                  (.getGroups (.getVersionNumber user-agent))))))))

(defn get-user-agent [request]
  (let [headers    (request :headers)
        user-agent (headers "user-agent")]
    (parse-user-agent-string user-agent)))

(defn new-enough? [family version]
  (cond (= "CHROME"  family) (>= version 25)
        (= "FIREFOX" family) (>= version 23)
        (= "OPERA"   family) (>= version 15)
        (= "OPERA_MOBILE" family) (>= version 14)
        :else false))

(defn standard-header? [browser]
  (let [browser-family    (browser :family)
        browser-version (first (browser :version))]
    (and (contains? #{"CHROME" "FIREFOX" "OPERA" "OPERA_MOBILE"} browser-family)
         (new-enough? browser-family browser-version))))

(defn gecko-header? [browser]
  (let [browser-family    (browser :family)
        browser-version (first (browser :version))]
    (or (and (= "FIREFOX" browser-family)
             (<= browser-version 23))
        (and (= "IE" browser-family)
             (<= browser-version 10))
        (= "MOZILLA" browser-family)))) ; IE 11 spoofs Mozilla!

(defn webkit-header? [browser]
  (let [browser-family  (browser :family)
        browser-version (first (browser :version))]
    (or (and (= "CHROME" browser-family)
             (<= browser-version 25))
        (and (= "CHROME_MOBILE" browser-family)
             (<= browser-version 28))
        (and (= "BLACKBERRY_BROWSER" browser-family)
             (<= 10 browser-version))
        (and (contains? #{"SAFARI" "MOBILE_SAFARI"} browser-family)
             (>= browser-version 5)))))

(defn select-header [request]
  (let [browser (get-user-agent request)]
    (cond (standard-header? browser)   "Content-Security-Policy"
          (gecko-header?    browser)   "X-Content-Security-Policy"
          (webkit-header?   browser)   "X-Webkit-CSP"
          :else "Content-Security-Policy")))
